src/libclipper/src/persistent_state.cpp

```diff
@@ -42,21 +48,58 @@ StateDB::StateDB() {
 StateDB::~StateDB() { redis_connection_.disconnect(); }
 
 boost::optional<std::string> StateDB::get(const StateKey& key) {
-  std::string redis_key = generate_redis_key(key);
-  const std::vector<std::string> cmd_vec{"GET", redis_key};
-  return redis::send_cmd_with_reply<std::string>(redis_connection_, cmd_vec);
+  auto entry_search = cache_.find(key);
+  if (entry_search != cache_.end()) {
+    auto& cache_entry = entry_search->second;
+    std::lock_guard<std::mutex> entry_lock(cache_entry->entry_mtx_);
+    return cache_entry->entry_value_;
+  } else {
+    std::string redis_key = generate_redis_key(key);
+    const std::vector<std::string> cmd_vec{"GET", redis_key};
+    auto result =
+        redis::send_cmd_with_reply<std::string>(redis_connection_, cmd_vec);
+    return result;
+  }
 }
 
 bool StateDB::put(StateKey key, std::string value) {
+  auto entry_search = cache_.find(key);
+  bool new_entry = (entry_search == cache_.end());
+  if (!new_entry) {
+    std::lock_guard<std::mutex> lock(entry_search->second->entry_mtx_);
+  }
   std::string redis_key = generate_redis_key(key);
   const std::vector<std::string> cmd_vec{"SET", redis_key, value};
-  return redis::send_cmd_no_reply<std::string>(redis_connection_, cmd_vec);
+  bool success =
+      redis::send_cmd_no_reply<std::string>(redis_connection_, cmd_vec);
+  if (success) {
+    if (new_entry) {
+      // This method is being invoked as part of the 'add application'
+      // procedure in the query frontend. Because any subsequent feedback
+      // updates to the StateDB depend on the completion of 'add application',
+      // it's safe to add a new cache entry immediately
```

Comment:
Hmm will this always be true? This seems like a complicated behavior to rely on for safety. What if another code path ends up putting things in the StateDB? Will that break this code?

Reply:
Ahh I understand your comment now. Can you clarify in the comment that *if we are creating a new entry, it means that the method is being invoked as part of adding an application*?

But my previous question still remains, this seems like tricky behavior to rely on for safety.